### PR TITLE
Add Authentication to the App

### LINKED ISSUE #3: Authentication Routes
**Authentication for the Users**
Strategy: JWT
Routes for:
- Login
- Signup
- Forgot Password
- Reset Password

### PR DESCRIPTION
**Pull request by sohamsshah**
This PR fixes #3 

Changes:
* Adds Auth routes to the backend
* routes added: `sign-up`, `sign-in` and `log-out`
* Deleted client folder because of the decision to migrate to `next` instead of `react`
* Added `ProtectedRoutes.ts` route to test the working of JWT. Can be removed later

Schemas added:
* UserCredentials
* UserDetails (refs -> UserCredentials)

Middlewares:
* AuthMiddleware
* ValidateMiddleware

Controllers:
* AuthController

### CODE REVIEW

File: src/validation/AuthValidation.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import * as yup from "yup";
+
+export const signUpSchema = yup.object({
+  username: yup.string().required("First name is a required field."),
+  firstName: yup.string().required("First name is a required field."),
+  lastName: yup.string().required("Last name is a required field."),
+  email: yup
+    .string()
+    .email("Please make sure entered email is valid.")
+    .required("Email is a required field"),
```
Comment (prerana1821) [inline src/validation/AuthValidation.ts:10]: Add email case-insensitive validation
```suggestion
  email: yup
    .string()
    .email("Please make sure entered email is valid.")
    .required("Email is a required field")
    .test(
         'existsCheck',
         'Email already exists',
         value => !lowerEmails.includes(value.toLowerCase())
      ),
```

File: src/validation/AuthValidation.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    .max(256)
+    .required("Password is a required field.")
+    .matches(
+      /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d@$!%*#?&]{8,}$/,
+      "Must Contain 8 Characters, mix of numbers and alphabets."
+    ),
+});
+
+export const signInSchema = yup.object({
+  email: yup
```
Comment (prerana1821) [inline src/validation/AuthValidation.ts:23]: Add email case-insensitive validation

File: src/controllers/AuthController.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+ * send POST Request at /api/auth/sign-up
+ * body contains {username, firstName, lastName, email, password}
+ * */
+export const signUpHandler: RequestHandler<{}, {}, SignUpBody> = async (
+  req,
+  res
+) => {
+  const { username, firstName, lastName, email, password } = req.body;
+
+  const isAlreadyRegistered = await UserCredentials.findOne({
```
Comment (prerana1821) [inline src/controllers/AuthController.ts:20]: Add this into try-catch

**Review by prerana1821 — CHANGES_REQUESTED** [review]
Hey @sohamsshah,
Please add the Request changes

File: src/validation/AuthValidation.ts (hunk 1 above)
Comment (sohamsshah) [reply src/validation/AuthValidation.ts:10]: what is lowerEmails in this case?

Comment (prerana1821) [reply src/validation/AuthValidation.ts:10]: email field

Comment (sohamsshah) [reply src/validation/AuthValidation.ts:10]: Yup was throwing error. We can handle this in upcoming PRs. Raising an issue for the same.

**Pull request merged**